### ISSUE TITLE
RetroArch PS3 graphics issues

### ISSUE THREAD

**Issue by twinaphex** [RPCS3/rpcs3#3905]
Hi there,

it seems that RPCS3 seems to have a lot of problems with file paths when I try to use RetroArch PS3 on the emulator. RetroArch PS3 was made with the real PS3 SDK instead of PSL1GHT. 

Version of RPCS3 used: **v0.0.4-6209-d8ae94df Alpha**

Further information provided - I use PSGL for graphics, I use stb_font for rendering the font glyphs, and I am using the runtime Cg compiler to load shaders on the fly from source.

Here is a log file -

https://hastebin.com/bosotizeca.js

Graphics problems:
* The icons don't show up, presumably because all the file system accesses are failing.
* The font seems to be corrupted.

This is what it looks like right now -

<img width="955" alt="retroarchps3" src="https://user-images.githubusercontent.com/1075274/33661579-d3e234be-da88-11e7-8856-b00478b683c7.png">

This is how it should look like instead (screenshot taken on PC with Cg shader backend) -

<img width="576" alt="retroarchps3-pc" src="https://user-images.githubusercontent.com/1075274/33661973-6405041c-da8a-11e7-85c6-eeeff2129dd7.png">

**How to test this**

Download the latest version of RetroArch for PS3 here -

https://buildbot.libretro.com/stable/1.6.9/playstation/ps3/

https://buildbot.libretro.com/stable/1.6.9/playstation/ps3/RetroArch.PS3.CEX.PS3.pkg

NOTE - It seems RPCS3 isn't yet capable of dealing with DEX debug SELFs. Tried using it but I couldn't get it to boot up so I had to resort to the CEX build instead.

Install the pkg in RPCS3. Start it up. It should auto default to one core out of the package.

**Source**

All of our source code for the PS3 port is publicly available on our repository -

https://github.com/libretro/RetroArch

Let me help you further along and point you to some of the PS3-specific code (beyond these files, all the normal codepaths that would normally get used on PC are used too here) -

**File I/O code specific to PS3 **

https://github.com/libretro/RetroArch/blob/master/frontend/drivers/platform_ps3.c

This file sets up all the filesystem paths - say, where to get the assets directory from, where to get the cores from, etc.

https://github.com/libretro/RetroArch/blob/master/libretro-common/file/retro_dirent.c

This file has some PS3-specific code for iterating over files.

**Graphics code specific to PS3 **

https://github.com/libretro/RetroArch/blob/master/gfx/drivers/gl.c

Basically the PS3 reuses the same GL driver as on PC. The only differences is that HAVE_PSGL, HAVE_OPENGLES are defined, so it employs some of the PSGL-optimized codepaths. For instance, it uses texture references which is some special construct available only in PSGL.

More PS3-specific code can be found in the renderchain file here -

https://github.com/libretro/RetroArch/blob/master/gfx/drivers_renderchain/gl2_renderchain.c

Hopefully this information was helpful, and hopefully it will help make RPCS3 an even better PS3 emulator! I'd be very glad if we could start using RPCS3 as a debugger for our PS3 port personally.

**Comment by clienthax** [comment]
Looks to be that we lose the APP_HOME vfs when the respawn is called or something

**Comment by clienthax** [comment]
On further checks, it looks like we need to traverse up the directory tree to find the .sfo file, instead of assuming we are one level down from it.

![image](https://user-images.githubusercontent.com/1056270/33664481-b0a44a5c-da8b-11e7-8dc4-c520de04511e.png)

**Comment by twinaphex** [comment]
Yeah that makes sense, since our SELF files are located in the directory 'cores' and the EBOOT.BIN is just a small exitspawn program that just 'spawns' any of those 'core' SELF files. So I guess it isn't laid out like normal PS3 games would.

Ah, looks like the textures are working. I don't know if that stable had wrong PNG images for those icons. Try to see if it happens with this file instead -

https://buildbot.libretro.com/nightly/playstation/ps3/RetroArch.PS3.CEX.PS3.pkg

These are the latest nightlies and should have the latest icon sets.

Seems like when this issue is resolved, the only issue remaining then is figuring out why the font renderer looks all messed up.

**Comment by clienthax** [comment]
@twinaphex Are you on discord btw, i see you are not on the irc i used to chat with you on, there also seems to be an issue with the autoupdater via rpcs3 that id like to chat about :)

**Comment by twinaphex** [comment]
Hi there, I am on #retroarch and #libretro on Freenode. I am also on the RetroArch discord but I use an IRC-to-Discord bridge there in order to talk.

**Comment by twinaphex** [comment]
My nick is DanteAli on there BTW.

**Comment by twinaphex** [comment]
Since it seems there are some issues with the texture formats, here are the PSGL texture formats that are being used -

https://github.com/libretro/RetroArch/blob/master/gfx/common/gl_common.h#L98

You would have to see what they correspond to on the libGCM side, since PSGL is just a GL-to-libgcm wrapper anyways.

**Comment by kd-11** [comment]
Texture channel problems are fixed by this commit https://github.com/kd-11/rpcs3/commit/94f1f4662b35484a33d151b752d1cdec242a9d9d but the traversal bug keeps the icons from loading in

![image](https://user-images.githubusercontent.com/15904127/33707495-95ea9b04-db48-11e7-94f1-db08063e4d53.png)

**Comment by clienthax** [comment]
Will test out on my build soon :)

**Comment by clienthax** [comment]
Hmm @kd-11 I seem to have a non transparent background here (very slightly)
![image](https://user-images.githubusercontent.com/1056270/33723404-872db04e-db64-11e7-9d0a-ccf2a2715c0e.png)

**Comment by twinaphex** [comment]
Yeah that non transparent background still seems wrong on the glyphs.

**Comment by kd-11** [comment]
Fixed in new commits on my fork

**Comment by AniLeo** [comment]
This has been fixed by #3914 #3931 and #3984 

![image](https://user-images.githubusercontent.com/10283761/34392647-64ef3ea8-eb45-11e7-8048-b75c74d9919e.png)

Closing as solved
